### PR TITLE
Add indexedDB + caching support

### PR DESCRIPTION
**Pull request by yusufhilmi**
This includes a refactor and couple of fixes on top of [@Lurrobert's PR](https://github.com/yusufhilmi/client-vector-search/pull/1)

There's still some tech debt we create with this PR but most indexedDB actions are at least abstracted in a way we can fix later with minimal breaking changes.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -142,33 +174,177 @@ export class EmbeddingIndex {
...
+        reject(new Error('Index is empty. Nothing to save'));
+        return;
+      }
+
+      const db = await IndexedDbManager.create(DBname, objectStoreName);
+
+      await db
+        .addToIndexedDB(this.objects)
+        .then(() => {
+          console.log(
```
Comment (yusufhilmi) [inline src/index.ts:284]: We probably dont want to log console anything unless its super critical or an error.

File: src/index.ts [hunk 2]
```diff
@@ -91,10 +123,10 @@ export class EmbeddingIndex {
   update(filter: { [key: string]: any }, vector: { [key: string]: any }) {
```
Comment (yusufhilmi) [inline src/index.ts:137]: we should have these types as interfaces and use them here and in other places.

File: src/index.ts [hunk 3]
```diff
@@ -1,32 +1,67 @@
...
+}
+
+const cacheInstance = Cache.getInstance();
+
 export const getEmbedding = async (
   text: string,
   precision: number = 7,
-  options = { pooling: "mean", normalize: false },
-  model = "Xenova/gte-small"
+  options = { pooling: 'mean', normalize: false },
```
Comment (Lurrobert) [inline src/index.ts:50]: update to the latest model initialization
```let transformersModule: any;
let pipeline: any;
let pipe: any;
let currentModel: string;

export const initializeModel = async (model: string = "Xenova/gte-small"): Promise<void> => {
  if (model !== currentModel) {
    transformersModule = await import("@xenova/transformers");
    pipeline = transformersModule.pipeline;
    pipe = await pipeline("feature-extraction", model);
    currentModel = model;
  }
}

export const getEmbedding = async (
  text: string,
  precision: number = 7,
  options = { pooling: "mean", normalize: false },
  model: string = "Xenova/gte-small"
): Promise<number[]> => {
  if (model !== currentModel) {
    await initializeModel(model);
  }...
```

Comment (yusufhilmi) [reply src/index.ts:50]: fixed now, we don't need to map pipeline and pipe, transformersModule is enough on global namespace

**Pull request merged**